include/aws/common/atomics_msvc.inc

```diff
@@ -0,0 +1,324 @@
+ * semantics as well, but because windows is generally used on x86, where there's not a lot
+ * of performance difference between different ordering modes anyway, we just use the stronger
+ * forms for now. Further, on x86, they actually have seq_cst semantics as they use locked instructions.
+ *
+ * Since all loads and stores are acq and/or rel already, we can do non-seq_cst loads and stores
+ * as just volatile variable accesses.
+ *
+ * For seq_cst accesses, we take advantage of the facts that:
+ * 1. Loads are not reordered with other loads
+ * 2. Stores are not reordered with other stores
+ * 3. Locked instructions (including swaps) have a total order
+ * 4. Non-locked accesses are not reordered with locked instructions
+ *
+ * Therefore, if we ensure that all seq_cst stores are locked, we can establish
+ * a total order on stores, and the intervening ordinary loads will not violate that total
+ * order.
+ * See http://www.cs.cmu.edu/~410-f10/doc/Intel_Reordering_318147.pdf 2.7, which covers
+ * this use case.
+ */
+
+/**
+ * Windows guarantees support for 64-bit ints in atomics.
+ */
+typedef int64_t aws_atomic_int_impl_t;
+
+#define AWS_ATOMIC_INT_IMPL_MAX INT64_MAX
+#define AWS_ATOMIC_INT_IMPL_MIN INT64_MIN
+
+#pragma warning(push)
+#pragma warning(disable : 4201) /* nameless union */
+
+struct aws_atomic_var {
+    union {
+        aws_atomic_int_impl_t intval;
+        void *ptrval;
+    };
+};
+
+#pragma warning(pop)
+
+static inline void aws_atomic_priv_check_order(enum aws_memory_order order) {
+#ifndef NDEBUG
+    switch (order) {
+        case aws_memory_order_relaxed:
+            return;
+        case aws_memory_order_acquire:
+            return;
+        case aws_memory_order_release:
+            return;
+        case aws_memory_order_acq_rel:
+            return;
+        case aws_memory_order_seq_cst:
+            return;
+        default: /* Unknown memory order */
+            abort();
+    }
+#endif
+    (void)order;
+}
+
+/**
+ * Initializes an atomic variable with an integer value. This operation should be done before any
+ * other operations on this atomic variable, and must be done before attempting any parallel operations.
+ */
+void aws_atomic_init_int(volatile struct aws_atomic_var *var, aws_atomic_int_impl_t n) {
+    var->intval = n;
+}
+
+/**
+ * Initializes an atomic variable with a pointer value. This operation should be done before any
+ * other operations on this atomic variable, and must be done before attempting any parallel operations.
+ */
+void aws_atomic_init_ptr(volatile struct aws_atomic_var *var, void *p) {
+    var->ptrval = p;
+}
+
+/**
+ * Reads an atomic var as an integer, using the specified ordering, and returns the result.
+ */
+AWS_STATIC_IMPL
+aws_atomic_int_impl_t aws_atomic_load_int(
+    volatile const struct aws_atomic_var *var,
+    enum aws_memory_order memory_order) {
+    aws_atomic_priv_check_order(memory_order);
+    return var->intval;
```

Comment:
I went debug-stepping through MSVC 2017's x64 code, they implemented it like this: 
```
_Uint8_t _Load_seq_cst_8(volatile _Uint8_t *_Tgt)
    _Uint8_t _Value;
    _Value = *_Tgt;
    _Compiler_barrier();
    return _Value;
```
(from xatomic.h)

Is it really safe to just read the value if Microsoft is using a barrier afterwards?

Reply:
https://msdn.microsoft.com/en-us/library/12a04hfd.aspx

A read of a volatile object (except on ARM, apparently) has acquire semantics. This means that no reads or writes that come after the read will be reordered prior to the read. This is stronger than the compiler barrier itself implies (as it includes hardware reordering as well). Most likely MSVC has that barrier for ARM support.

Reply:
This is correct, seq_cst only needs hardware barriers on stores on x86.